contrib/kafka-humio-gateway/main.go

```diff
@@ -0,0 +1,288 @@
+
+	sigterm := make(chan os.Signal, 1)
+	signal.Notify(sigterm, syscall.SIGINT, syscall.SIGTERM)
+
+	for keepRunning {
+		select {
+		case <-ctx.Done():
+			log.Println("terminating: context cancelled")
+			keepRunning = false
+		case <-sigterm:
+			log.Println("terminating: via signal")
+			keepRunning = false
+		}
+	}
+	cancel()
+	wg.Wait()
+	if err = client.Close(); err != nil {
+		log.Panicf("Error closing client: %v", err)
+	}
+}
+
+// Consumer represents a Sarama consumer group consumer
+type Consumer struct {
+	ready chan bool
+	httpClient http.Client
+	config TransportConfig
+	performTransformations bool
+}
+
+// Setup is run at the beginning of a new session, before ConsumeClaim
+func (consumer *Consumer) Setup(sarama.ConsumerGroupSession) error {
+	// Mark the consumer as ready
+	close(consumer.ready)
+	return nil
+}
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+func (consumer *Consumer) Cleanup(sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+
+// The Row data can be string or int.  We don't care which.  We just pass it through.
+type VRRKafkaMessage struct {
+        Tags       map[string]interface{}    `json:"tags"`
+        RowData    map[string]interface{}    `json:"row_data"`
+        Version    int                       `json:"version"`
+}
+
+type HumioEvent struct {
+	// Could be string or int.  We don't need to care which.
+        Timestamp interface{}               `json:"timestamp"`
+        Attributes  map[string]interface{}  `json:"attributes"`
+        Timezone string                     `json:"timezone,omitempty"`
+}
+
+type HumioPayload struct {
+        Events []HumioEvent                 `json:"events"`
+        Tags map[string]interface{}         `json:"tags,omitempty"`
+}
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+func (consumer *Consumer) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	// NOTE:
+	// Do not move the code below to a goroutine.
+	// The `ConsumeClaim` itself is called within a goroutine, see:
+	// https://github.com/Shopify/sarama/blob/main/consumer_group.go#L27-L29
+	for message := range claim.Messages() {
+		var data []byte
+		var err error
+		if verbose {
+			log.Printf("Received message Topic[%s] Key[%s] Value[%s] Timestamp[%v]", message.Topic, message.Key, message.Value, message.Timestamp)
+		}
+
+		var values VRRKafkaMessage
+		err = json.Unmarshal(message.Value, &values)
+		if err != nil {
+			log.Printf("Failed to unmarshal [%s]: %v", message.Value, err)
+			session.MarkMessage(message, "")
+			continue
+		}
+
+		timestamp, ok := values.RowData["Time"]
+		timezone := ""
+		if !ok {
+			timestamp, ok = values.RowData["timestamp"]
+			timezone = "UTC"
+		}
+
+
+		if !ok {
+			log.Printf("Failed to gather timestamp from event: %s", message.Value)
+			session.MarkMessage(message, "")
+			continue
+		}
+
+		payload := HumioPayload{
+			Events: []HumioEvent {
```

Comment:
Here we only ever have a single event?

Reply:
Yes, but the API for Humio requires that the events be submitted in a JSON array. Eventually we could submit multiple events in a single API call, but that would require that all events in the same submission use the same tags. It's unclear how we want to use tags generally, but for now we're using hostname, client id, and VRR artifact. We could batch them but I'm not sure how much benefit that would bring compared with the complexity.

Reply:
Fair enough.

Reply:
I'm going to revisit this. As I started enabling noisier artifacts like process execution and network activity, it became obvious we will need to implement some kind of batching here.

Reply:
I've revisited this in commit [b8be7d22](https://github.com/SUSE/linux-security-sensor/pull/8/commits/b8be7d22b7c76f375a8c73ddcb6b55e2ed957f63).